vm/devices/storage/disk_ramdisk/src/lib.rs

```diff
@@ -69,45 +68,35 @@ pub enum Error {
     /// Invalid disk geometry.
     #[error(transparent)]
     InvalidGeometry(#[from] InvalidGeometry),
-    /// Unsupported sector size.
-    #[error("unsupported sector size {0}")]
-    UnsupportedSectorSize(u32),
 }
 
 struct Sector([u8; 512]);
 
 const SECTOR_SIZE: u32 = 512;
 
-impl RamDisk {
+impl RamLayer {
     /// Makes a new RAM disk of `size` bytes.
-    pub fn new(len: u64, read_only: bool) -> Result<Self, Error> {
-        Self::new_inner(
-            Disk::new(ZeroDisk::new(SECTOR_SIZE, len)?).unwrap(),
-            read_only,
-            true,
-        )
-    }
-
-    /// Makes a new RAM diff disk on top of `lower`.
     ///
-    /// Writes will be collected in RAM, but reads will go to the lower disk for
-    /// sectors that have not yet been overwritten.
-    pub fn diff(lower: Disk, read_only: bool) -> Result<Self, Error> {
-        Self::new_inner(lower, read_only, false)
-    }
-
-    fn new_inner(lower: Disk, read_only: bool, lower_is_zero: bool) -> Result<Self, Error> {
-        let sector_size = lower.sector_size();
-        if sector_size != SECTOR_SIZE {
-            return Err(Error::UnsupportedSectorSize(sector_size));
-        }
-        let sector_count = lower.sector_count();
+    /// If `None` is specified, then the disk will inherit its size from the
+    /// lower layer when it is attached to a [`LayeredDisk`].
+    pub fn new(size: Option<u64>) -> Result<Self, Error> {
+        let sector_count = if let Some(size) = size {
+            if size == 0 {
+                return Err(Error::InvalidGeometry(InvalidGeometry::EmptyDisk));
+            }
+            if size % SECTOR_SIZE as u64 != 0 {
+                return Err(Error::InvalidGeometry(InvalidGeometry::NotSectorMultiple {
+                    disk_size: size,
+                    sector_size: SECTOR_SIZE,
+                }));
+            }
+            size / SECTOR_SIZE as u64
+        } else {
+            0
```

Comment:
preference for using an `Option<AtomicU64>` here instead of using 0 as a sentinel, such that downstream code needs to be a bit more thoughtful on handling operations on leyrs that aren't actually attached to anything yet

Reply:
Ehh, conceptually the disk just has zero sectors until it's attached, and attaching it implicitly resizes it. I think this is not unreasonable. The `Option` adds annoying complexity elsewhere.

Reply:
Ok